File: Examples/GroupDocs.Annotation.VisualBasic/GroupDocs.AnnotationExamples/CommonUtilities.vb
Imports GroupDocs.Annotation.Config
Imports GroupDocs.Annotation.Domain
Imports GroupDocs.Annotation.Domain.Image
Imports GroupDocs.Annotation.Domain.Options
Imports GroupDocs.Annotation.Handler
Imports GroupDocs.Annotation.Handler.Input
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

'ExStart:CommonUtilities
Class CommonUtilities
    'ExStart:CommonProperties
    Private Const StorageFolderPath As String = "../../../../Data/Samples/"
    Private Const DestinationFolderPath As String = "../../../../Data/Output/"
    Private Const LicenseFilePath As String = "D://Groupdocs.Total.lic"
    'ExEnd:CommonProperties

    'ExStart:MapSourceFilePath
    ''' <summary>
    ''' Maps source file path
    ''' </summary>
    ''' <param name="SourceFileName">Source File Name</param>
    ''' <returns>Returns complete path of source file</returns>
    Public Shared Function MapSourceFilePath(SourceFileName As String) As String
        Try
            Return StorageFolderPath & SourceFileName
        Catch exp As System.Exception
            Console.WriteLine(exp.Message)
            Return exp.Message
        End Try
    End Function
    'ExEnd:MapSourceFilePath
    'ExStart:MapDestinationFilePath
    ''' <summary>
    ''' Maps destination file path
    ''' </summary>
    ''' <param name="DestinationFileName">Destination File Name</param>
    ''' <returns>Returns complete path of destination file</returns>
    Public Shared Function MapDestinationFilePath(DestinationFileName As String) As String
        Try
            Return DestinationFolderPath & DestinationFileName
        Catch exp As System.Exception
            Console.WriteLine(exp.Message)
            Return exp.Message
        End Try
    End Function
    'ExEnd:MapDestinationFilePath

    ''' <summary>
    ''' Saves output document
    ''' </summary>

    Public Shared Sub SaveOutputDocument(inputFile As Stream, annotations As List(Of AnnotationInfo), type As DocumentType)
        Try
            'ExStart:SaveOutputDocument
            ' Create instance of annotator. 
            Dim cfg As AnnotationConfig = CommonUtilities.GetConfiguration()

            Dim annotator As New AnnotationImageHandler(cfg)

            Dim documentRepository As IDocumentDataHandler = annotator.GetDocumentDataHandler()
            If Not Directory.Exists(cfg.StoragePath) Then
                Directory.CreateDirectory(cfg.StoragePath)
            End If

            Dim result As Stream = annotator.ExportAnnotationsToDocument(inputFile, annotations, type)

            Dim getFileStream As FileStream = TryCast(inputFile, FileStream)
            Dim extensionWithDot As String = Path.GetExtension(getFileStream.Name)

            ' Save result stream to file.
            Using fileStream As New FileStream(MapDestinationFilePath(Convert.ToString("Annotated") & extensionWithDot), FileMode.Create)
                Dim buffer As Byte() = New Byte(result.Length - 1) {}
                result.Seek(0, SeekOrigin.Begin)
                result.Read(buffer, 0, buffer.Length)
                fileStream.Write(buffer, 0, buffer.Length)
                fileStream.Close()
                'ExEnd:SaveOutputDocument
            End Using
        Catch exp As System.Exception
            Console.WriteLine(exp.Message)
        End Try
    End Sub


    'ExStart:ApplyLicense
    ''' <summary>
    ''' Applies product license
    ''' </summary>
    Public Shared Sub ApplyLicense()
        Try
            ' initialize License
            Dim lic As New License()
            ' apply license
            lic.SetLicense(LicenseFilePath)
        Catch exp As System.Exception
            Console.WriteLine(exp.Message)
        End Try
    End Sub
    'ExEnd:ApplyLicense

    'ExStart:GetConfiguration
    ''' <summary>
    ''' Sets annotation configuration
    ''' </summary>
    ''' <returns>Returns AnnotationConfig object</returns>
    Public Shared Function GetConfiguration() As AnnotationConfig
        Try
            Dim cfg As New AnnotationConfig()

            'Set storage path 
            cfg.StoragePath = StorageFolderPath

            Return cfg
        Catch exp As System.Exception
            Console.WriteLine(exp.Message)
            Return Nothing
        End Try
    End Function
    'ExEnd:GetConfiguration

    'ExStart:GetImageRepresentation
    ''' <summary>
    ''' Gets image representation of document
    ''' </summary>
    ''' <param name="filePath">Source file path</param> 
    Public Shared Sub GetImageRepresentation(filePath As String)
        Try
            Dim document As Stream = New FileStream(MapSourceFilePath(filePath), FileMode.Open)
            Dim cfg As AnnotationConfig = GetConfiguration()

            Dim annotationHandler As New AnnotationImageHandler(cfg)

            Dim images As List(Of PageImage) = annotationHandler.GetPages(document, New ImageOptions())

            ' Save result stream to file.
            Using fileStream As New FileStream(MapDestinationFilePath("image.png"), FileMode.Create)
                Dim buffer As Byte() = New Byte(images(0).Stream.Length - 1) {}
                images(0).Stream.Seek(0, SeekOrigin.Begin)
                images(0).Stream.Read(buffer, 0, buffer.Length)
                fileStream.Write(buffer, 0, buffer.Length)
                fileStream.Close()
            End Using
        Catch exp As System.Exception
            Console.WriteLine(exp.Message)
        End Try

    End Sub
    'ExEnd:GetImageRepresentation

    'ExStart:GetTextCoordinatesInImage
    ''' <summary>
    ''' Gets text coordinates in image representation of document
    ''' </summary>
    ''' <param name="filePath">Source file path</param> 
    Public Shared Sub GetTextCoordinates(filePath As String)
        Try
            ' Set configuration
            Dim cfg As AnnotationConfig = GetConfiguration()

            ' Initialize annotator 
            Dim annotator As New AnnotationImageHandler(cfg)
            Try
                annotator.CreateDocument(filePath)
            Catch
            End Try

            Dim documentInfoContainer = annotator.GetDocumentInfo(filePath)

            ' Go through all pages
            For Each pageData As PageData In documentInfoContainer.Pages
                Console.WriteLine("Page number: " + pageData.Number)

                'Go through all page rows
                For i As Integer = 0 To pageData.Rows.Count - 1
                    Dim rowData As RowData = pageData.Rows(i)

                    ' Write data to console
                    Console.WriteLine("Row: " + (i + 1))
                    Console.WriteLine("Text: " + rowData.Text)
                    Console.WriteLine("Text width: " + rowData.LineWidth)
                    Console.WriteLine("Text height: " + rowData.LineHeight)
                    Console.WriteLine("Distance from left: " + rowData.LineLeft)
                    Console.WriteLine("Distance from top: " + rowData.LineTop)

                    ' Get words
                    Dim words As String() = rowData.Text.Split(" "c)

                    ' Go through all word coordinates
                    For j As Integer = 0 To words.Length - 1
                        Dim coordinateIndex As Integer = If(j = 0, 0, j + 1)
                        ' Write data to console
                        Console.WriteLine(String.Empty)
                        Console.WriteLine("Word:'" + words(j) + "'")
                        Console.WriteLine("Word distance from left: " + rowData.TextCoordinates(coordinateIndex))
                        Console.WriteLine("Word width: " + rowData.TextCoordinates(coordinateIndex + 1))
                        Console.WriteLine(String.Empty)
                    Next
                    Console.ReadKey()
                Next
            Next
        Catch exp As System.Exception
            Console.WriteLine(exp.Message)
        End Try

    End Sub
    'ExEnd:GetTextCoordinatesInImage
End Class
'ExEnd:CommonUtilities 

File: Examples/GroupDocs.Annotation.Examples.CSharp/GroupDocs.AnnotationExamples/CommonUtilities.cs
using GroupDocs.Annotation.Common.License;
using GroupDocs.Annotation.Config;
using GroupDocs.Annotation.Domain;
using GroupDocs.Annotation.Domain.Image;
using GroupDocs.Annotation.Domain.Options;
using GroupDocs.Annotation.Handler;
using GroupDocs.Annotation.Handler.Input;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GroupDocs.Annotation.CSharp
{
    //ExStart:CommonUtilities
    class CommonUtilities
    {
        //ExStart:CommonProperties
        public static string StorageFolderPath = "../../../../Data/Samples/";
        public static string DestinationFolderPath = "../../../../Data/Output/";
        public static string LicenseFilePath = "E://License/Groupdocs.Total.lic";
        public static string filePath = "Annotated.pdf";
        //ExEnd:CommonProperties



        //ExStart:MapSourceFilePath
        /// <summary>
        /// Maps source file path
        /// </summary>
        /// <param name="SourceFileName">Source File Name</param>
        /// <returns>Returns complete path of source file</returns>
        public static string MapSourceFilePath(string SourceFileName)
        {
            try
            {
                return StorageFolderPath + SourceFileName;
            }
            catch (System.Exception exp)
            {
                Console.WriteLine(exp.Message);
                return exp.Message;
            }
        }
        //ExEnd:MapSourceFilePath
        //ExStart:MapDestinationFilePath
        /// <summary>
        /// Maps destination file path
        /// </summary>
        /// <param name="DestinationFileName">Destination File Name</param>
        /// <returns>Returns complete path of destination file</returns>
        public static string MapDestinationFilePath(string DestinationFileName)
        {
            try
            {
                return DestinationFolderPath + DestinationFileName;
            }
            catch (System.Exception exp)
            {
                Console.WriteLine(exp.Message);
                return exp.Message;
            }
        }
        //ExEnd:MapDestinationFilePath

        /// <summary>
        /// Saves output document
        /// </summary>
        public static void SaveOutputDocument(Stream inputFile, List<AnnotationInfo> annotations, DocumentType type)
        {
            try
            {
                //ExStart:SaveOutputDocument
                // Create instance of annotator. 
                AnnotationConfig cfg = CommonUtilities.GetConfiguration();

                AnnotationImageHandler annotator = new AnnotationImageHandler(cfg);

                IDocumentDataHandler documentRepository = annotator.GetDocumentDataHandler();
                if (!Directory.Exists(cfg.StoragePath))
                {
                    Directory.CreateDirectory(cfg.StoragePath);
                }

                Stream result = annotator.ExportAnnotationsToDocument(inputFile, annotations, type);

                FileStream getFileStream = inputFile as FileStream;
                string extensionWithDot = Path.GetExtension(getFileStream.Name);

                // Save result stream to file.
                using (FileStream fileStream = new FileStream(MapDestinationFilePath("Annotated" + extensionWithDot), FileMode.Create))
                {
                    byte[] buffer = new byte[result.Length];
                    result.Seek(0, SeekOrigin.Begin);
                    result.Read(buffer, 0, buffer.Length);
                    fileStream.Write(buffer, 0, buffer.Length);
                    fileStream.Close();
                }
                //ExEnd:SaveOutputDocument
            }
            catch (System.Exception exp)
            {
                Console.WriteLine(exp.Message);
            }
        }

        //ExStart:ApplyLicense
        /// <summary>
        /// Applies product license
        /// </summary>
        public static void ApplyLicense()
        {
            try
            {
                // initialize License
                License lic = new License();
                // apply license
                lic.SetLicense(LicenseFilePath);
            }
            catch (System.Exception exp)
            {
                Console.WriteLine(exp.Message);
            }
        }
        //ExEnd:ApplyLicense

        //ExStart:GetConfiguration
        /// <summary>
        /// Sets annotation configuration
        /// </summary>
        /// <returns>Returns AnnotationConfig object</returns>
        public static AnnotationConfig GetConfiguration()
        {
            try
            {
                AnnotationConfig cfg = new AnnotationConfig();

                //Set storage path 
                cfg.StoragePath = StorageFolderPath;

                return cfg;
            }
            catch (System.Exception exp)
            {
                Console.WriteLine(exp.Message);
                return null;
            }
        }
        //ExEnd:GetConfiguration

        //ExStart:GetImageRepresentation
        /// <summary>
        /// Gets image representation of document
        /// </summary>
        /// <param name="CommonUtilities.filePath">Source file path</param> 
        public static void GetImageRepresentation(string filePath)
        {
            try
            {
                Stream document = new FileStream(MapSourceFilePath(filePath), FileMode.Open);
                AnnotationConfig cfg = GetConfiguration();

                AnnotationImageHandler annotationHandler = new AnnotationImageHandler(cfg);

                List<PageImage> images = annotationHandler.GetPages(document, new ImageOptions());

                // Save result stream to file.
                using (FileStream fileStream = new FileStream(MapDestinationFilePath("image.png"), FileMode.Create))
                {
                    byte[] buffer = new byte[images[0].Stream.Length];
                    images[0].Stream.Seek(0, SeekOrigin.Begin);
                    images[0].Stream.Read(buffer, 0, buffer.Length);
                    fileStream.Write(buffer, 0, buffer.Length);
                    fileStream.Close();
                }
            }
            catch (System.Exception exp)
            {
                Console.WriteLine(exp.Message);
            }

        }
        //ExEnd:GetImageRepresentation

        //ExStart:GetTextCoordinatesInImage
        /// <summary>
        /// Gets text coordinates in image representation of document
        /// </summary>
        /// <param name="CommonUtilities.filePath">Source file path</param> 
        public static void GetTextCoordinates(string filePath)
        {
            try
            {
                // Set configuration
                AnnotationConfig cfg = GetConfiguration();

                // Initialize annotator 
                AnnotationImageHandler annotator = new AnnotationImageHandler(cfg);
                try
                {
                    annotator.CreateDocument(filePath);
                }
                catch { }

                var documentInfoContainer = annotator.GetDocumentInfo(filePath);

                // Go through all pages
                foreach (PageData pageData in documentInfoContainer.Pages)
                {
                    Console.WriteLine("Page number: " + pageData.Number);

                    //Go through all page rows
                    for (int i = 0; i < pageData.Rows.Count; i++)
                    {
                        RowData rowData = pageData.Rows[i];

                        // Write data to console
                        Console.WriteLine("Row: " + (i + 1));
                        Console.WriteLine("Text: " + rowData.Text);
                        Console.WriteLine("Text width: " + rowData.LineWidth);
                        Console.WriteLine("Text height: " + rowData.LineHeight);
                        Console.WriteLine("Distance from left: " + rowData.LineLeft);
                        Console.WriteLine("Distance from top: " + rowData.LineTop);

                        // Get words
                        string[] words = rowData.Text.Split(' ');

                        // Go through all word coordinates
                        for (int j = 0; j < words.Length; j++)
                        {
                            int coordinateIndex = j == 0 ? 0 : j + 1;
                            // Write data to console
                            Console.WriteLine(string.Empty);
                            Console.WriteLine("Word:'" + words[j] + "'");
                            Console.WriteLine("Word distance from left: " + rowData.TextCoordinates[coordinateIndex]);
                            Console.WriteLine("Word width: " + rowData.TextCoordinates[coordinateIndex + 1]);
                            Console.WriteLine(string.Empty);
                        }
                        Console.ReadKey();
                    }
                }
            }
            catch (System.Exception exp)
            {
                Console.WriteLine(exp.Message);
            }

        }
        //ExEnd:GetTextCoordinatesInImage
    }
    //ExEnd:CommonUtilities
}
